### PR TITLE
feat: grafana agent

### PR DESCRIPTION
**Pull request by yanksyoon**
Applicable spec: N/A

### Overview

- Add grafana agent integration according to https://charmhub.io/topics/canonical-observability-stack/tutorials/instrumenting-machine-charms#step-4-relate-grafana-agent-to-cos-lite-prometheus-loki-and-grafana
- Charm resource detect & fix.

### Rationale

- To have log observability into the charm.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [ ] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

The documentation to be updated on a separate PR for easier PR processing.

### CODE REVIEW

File: src/cos.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    """The Jenkins COS integration observer."""
+
+    def __init__(self, charm: ops.CharmBase):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "cos-observer")
+        self._grafana_agent = COSAgentProvider(self)
```
Comment (cbartz) [inline src/cos.py:21]: Note that per default the logs from `/var/log/**/*log"` are transferred to Loki.

**Review by cbartz — DISMISSED** [review]
LGTM

**Review by jdkandersson — DISMISSED** [review]

File: tests/unit/test_state.py [hunk 2]
```diff
@@ -230,13 +231,26 @@ def test_builder_run_config(monkeypatch: pytest.MonkeyPatch):
...
         ),
         interval=6,
         unit_name=charm.unit.name,
     )
     assert result.run_config.cloud_config.cloud_name == state.CLOUD_NAME
 
 
+def test__get_num_parallel_build_error(monkeypatch: pytest.MonkeyPatch):
+    """
+    arrange: given monkeypatched multiprocessing.cpu_count() function thatt returns 1 core.
```
Comment (cbartz) [inline tests/unit/test_state.py:244]:
```suggestion
    arrange: given monkeypatched multiprocessing.cpu_count() function that returns 1 core.
```

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -35,7 +35,26 @@ async def test_image_relation(app: Application, test_charm: Application):
...
+    model: Model = app.model
+    grafana_agent = await model.deploy(
+        "grafana-agent",
+        application_name=f"grafana-agent-{app.name}",
+        channel="latest/edge",
+    )
+    await model.relate(f"{app.name}:cos-agent", f"{grafana_agent.name}:cos-agent")
+    await model.wait_for_idle(apps=[app.name], status="active", timeout=30 * 60)
+    # limit private-endpoint testing to machine relation test due to resource constraints of
+    # deploying k8s stack.
```
Comment (cbartz) [inline tests/integration/test_charm.py:57]: I think the comment is not necessary, or could you explain the alternative test scenario (I guess that logs are really sent to loki, but I think this is overkill for an integration test and also looks more like an end-to-end test)?

**Review by cbartz — APPROVED** [review]
LGTM

File: src/charm.py [hunk 4]
```diff
@@ -36,6 +37,7 @@ def __init__(self, *args: typing.Any):
         self.on.define_event("run", RunEvent)
 
         self.image_observer = image.Observer(self)
+        self.cos_observer = cos.Observer(self)
```
Comment (Thanhphan1147) [inline src/charm.py:40]: Is there a specific reason on why we don't directly instantiate the cosAgent class here but instead make a wrapper class for it?

File: src/state.py
Comment (Thanhphan1147) [inline src/state.py:1]: This doesn't seem to be related to the PR, is it intended here?

**Review by Thanhphan1147 — APPROVED** [review]
LGTM, just a few comments

File: src/state.py
Comment (yanksyoon) [reply src/state.py:1]: This is a follow up from the previous PR that @cbartz has left :D

File: src/cos.py [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+"""Observer module for Jenkins to COS integration."""
+
+
+import ops
+from charms.grafana_agent.v0.cos_agent import COSAgentProvider
+
+
+class Observer(ops.Object):
+    """The Jenkins COS integration observer."""
```
Comment (cbartz) [inline src/cos.py:12]:
```suggestion
    """The COS integration observer."""
```

File: tests/integration/test_charm.py (hunk 3 above)
Comment (yanksyoon) [reply tests/integration/test_charm.py:57]: Removed comment, thank you!
